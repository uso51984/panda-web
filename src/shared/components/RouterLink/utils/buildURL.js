import isObject from 'lodash/isObject';
import { location } from '@/shared/services/location';

export default to => {
  let path;
  if (isObject(to)) {
    const { url, params, search } = to;
    path = location.buildURL(url, { params, search });
  } else {
    path = location.buildURL(to);
  }

  return path;
};
