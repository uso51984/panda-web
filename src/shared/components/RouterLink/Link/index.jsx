import React from 'react';
import { Link } from 'react-router-dom';
import buildURL from '../utils/buildURL';

export default ({ children, to, ...rest }) => (<Link to={buildURL(to)} {...rest}>{children}</Link>);
