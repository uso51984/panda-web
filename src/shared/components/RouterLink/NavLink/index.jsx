import React from 'react';
import { NavLink } from 'react-router-dom';
import buildURL from '../utils/buildURL';

export default ({ children, to, ...rest }) =>
  (<NavLink to={buildURL(to)} {...rest}>{children}</NavLink>);
