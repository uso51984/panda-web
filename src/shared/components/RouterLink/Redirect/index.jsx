import React from 'react';
import { Redirect } from 'react-router-dom';
import buildURL from '../utils/buildURL';

export default ({ to, ...rest }) => (<Redirect to={buildURL(to)} {...rest} />);
