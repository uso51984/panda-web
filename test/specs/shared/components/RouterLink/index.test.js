
import React from 'react';
import renderer from 'react-test-renderer';
import Link from '@/shared/components/RouterLink/Link';
import NavLink from '@/shared/components/RouterLink/NavLink';
import Redirect from '@/shared/components/RouterLink/Redirect';
import buildURL from '@/shared/components/RouterLink/utils/buildURL';


jest.mock('react-router-dom', () => {
  const LinkNode = ({ to, children, ...rest }) => <a href={to} {...rest}>{children}</a>;

  return ({
    Link: LinkNode,
    NavLink: LinkNode,
    Redirect: LinkNode,
  });
});


describe('RouterLink', () => {
  it('buildURL', () => {
    expect(buildURL('http://www.facebook.com')).toEqual('http://www.facebook.com?locale=zh-CN');
    expect(buildURL({
      url: '/user/:id/name/:age',
      params: {
        id: 23,
        age: 25,
      },
      search: {
        name: 'chenjianbin',
      },
    })).toEqual('/user/23/name/25?name=chenjianbin&locale=zh-CN');
  });

  it('Link renders correctly', () => {
    const tree = renderer
      .create(<Link to="http://www.facebook.com" className="test">Facebook</Link>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('NavLink renders correctly', () => {
    const tree = renderer
      .create(<NavLink to="http://www.facebook.com" className="test">Facebook</NavLink>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Redirect renders correctly', () => {
    const tree = renderer
      .create(<Redirect to="http://www.facebook.com" className="test">Facebook</Redirect>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});


