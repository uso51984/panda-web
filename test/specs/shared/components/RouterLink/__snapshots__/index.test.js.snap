// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouterLink Link renders correctly 1`] = `
<a
  className="test"
  href="/user/23/name/25?locale=zh-CN"
>
  Facebook
</a>
`;

exports[`RouterLink NavLink renders correctly 1`] = `
<a
  className="test"
  href="/user/23/name/25?locale=zh-CN"
>
  Facebook
</a>
`;

exports[`RouterLink Redirect renders correctly 1`] = `
<a
  className="test"
  href="/user/23/name/25?locale=zh-CN"
>
  Facebook
</a>
`;
